### PR TITLE
Improve the bytecode interpreter

### PR DESCRIPTION
**Pull request by silene**
This pull request might seem large, but all the commits are straightforward. They are cleanups and optimizations of the bytecode interpreter.

Depends on #12894

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Queued bench https://gitlab.com/coq/coq/-/jobs/770281333

File: kernel/vmbytegen.ml
Comment (ppedrot) [inline kernel/vmbytegen.ml:663]: If we want to merge the native and the VM ASTs, it might be better to keep this dead code.

Comment (silene) [reply kernel/vmbytegen.ml:663]: If I am not mistaken, no code of `nativelambda.ml` can create `Lif` constructs. In other words, even after merging both native and bytecode ASTs, this part of the code would still stay dead. (This does not mean that the corresponding code in `nativecode.ml` is dead too. Indeed, contrary to `Lif`, the `MLif` construct can be generated, but that is specific to the native backend.)

**Comment by silene** [comment]
The timing differences seem to be just noise:

```
┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬────────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]      │    mem faults     │
│                             │                         │                                             │                                             │                                │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD  PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-bignums │   57.21   57.75 -0.94 % │      158687627138      159298527134 -0.38 % │      213073642148      213057650266 +0.01 % │     483160     471768  +2.41 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-verdi │  103.00  103.44 -0.43 % │      283864758330      284236366496 -0.13 % │      398344424703      398370832142 -0.01 % │     548396     548512  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1458.96 1462.72 -0.26 % │     4072114321044     4078173142623 -0.15 % │     6197817808365     6197761856040 +0.00 % │    1186552    1187432  -0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   47.46   47.58 -0.25 % │      131806823813      132070663332 -0.20 % │      181319518809      181321059635 -0.00 % │     486212     486496  -0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   49.97   50.07 -0.20 % │      137131314191      136977600534 +0.11 % │      173730708740      173700297423 +0.02 % │     524872     525044  -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  646.33  647.55 -0.19 % │     1810914866002     1812744206460 -0.10 % │     2736403725073     2736658907258 -0.01 % │    2987432    2996864  -0.31 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  104.60  104.78 -0.17 % │      299179258039      298922815657 +0.09 % │      398129700113      398159668826 -0.01 % │     487612     486724  +0.18 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-engine-bench-lite │  380.16  380.50 -0.09 % │     1070707388080     1069811877619 +0.08 % │     1836015065298     1836299408837 -0.02 % │    3974376    3507224 +13.32 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  226.56  226.61 -0.02 % │      630884632827      630036666862 +0.13 % │      959350893985      959290064344 +0.01 % │     695256     695468  -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  171.52  171.54 -0.01 % │      477397724873      476726308408 +0.14 % │      672246605677      672223573428 +0.00 % │     690828     686964  +0.56 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2554.54 2554.35 +0.01 % │     7121906939939     7120260621329 +0.02 % │    11807899681622    11796681076683 +0.10 % │     773884     774876  -0.13 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  663.14  662.77 +0.06 % │     1848213636217     1845392184685 +0.15 % │     2554803061877     2554549198930 +0.01 % │    1129276    1130032  -0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                         coq │  527.97  527.60 +0.07 % │     1478459779776     1475611655320 +0.19 % │     1993141580083     1993293592584 -0.01 % │     596300     596716  -0.07 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-coquelicot │   69.37   69.31 +0.09 % │      190075082494      189993952004 +0.04 % │      242910714876      242953993818 -0.02 % │     585852     584264  +0.27 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│  coq-performance-tests-lite │  544.22  543.68 +0.10 % │     1519988410519     1519173113847 +0.05 % │     2792384786294     2791173550333 +0.04 % │     749792     751264  -0.20 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-coqutil │   60.39   60.31 +0.13 % │      168554504798      167855004446 +0.42 % │      208891022848      208848336954 +0.02 % │     530668     530852  -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  533.04  532.02 +0.19 % │     1488366108479     1486395829294 +0.13 % │     1894533211381     1893654716737 +0.05 % │    1421544    1421824  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 4991.74 4981.20 +0.21 % │    13869597161828    13838322748371 +0.23 % │    24019066713153    23987672775780 +0.13 % │    2435216    2434228  +0.04 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  156.89  156.54 +0.22 % │      435421078630      435000729402 +0.10 % │      601238560746      601318125772 -0.01 % │     717944     749828  -4.25 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  214.06  213.57 +0.23 % │      597097523855      596158763846 +0.16 % │      825801680654      825778735382 +0.00 % │     518304     515492  +0.55 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-rewriter │  608.46  606.82 +0.27 % │     1688241663449     1683904444924 +0.26 % │     2576416520182     2576497690979 -0.00 % │    1034680    1036856  -0.21 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1249.52 1245.45 +0.33 % │     3476713926159     3465750194804 +0.32 % │     4954675000255     4950853386692 +0.08 % │    1171080    1170024  +0.09 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1240.44 1236.00 +0.36 % │     3452409760833     3439974760697 +0.36 % │     4877279270412     4877658936523 -0.01 % │     932916     932816  +0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  140.61  140.08 +0.38 % │      390337932700      389281531959 +0.27 % │      499729415936      499673759498 +0.01 % │     565000     561196  +0.68 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-unimath │ 4016.00 3998.51 +0.44 % │    11173551974810    11122674180153 +0.46 % │    21162641498611    21164472682559 -0.01 % │    1117900    1116800  +0.10 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-corn │ 1636.96 1629.44 +0.46 % │     4565059292289     4537084661307 +0.62 % │     6996374340995     6996389343864 -0.00 % │     831016     831124  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-hott │  325.33  323.65 +0.52 % │      880560783701      878092498435 +0.28 % │     1425605562868     1425372661986 +0.02 % │     726272     733960  -1.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-flocq │  158.73  157.91 +0.52 % │      440694036160      439143332859 +0.35 % │      549224763305      549215515821 +0.00 % │     820140     821992  -0.23 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1065.06 1059.16 +0.56 % │     2967250117012     2948089542909 +0.65 % │     4720141245748     4719703083024 +0.01 % │    1007124    1005236  +0.19 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  225.84  224.54 +0.58 % │      624570891235      621904408412 +0.43 % │     1027992301630     1027324833866 +0.06 % │    1084168    1087124  -0.27 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1046.90 1040.51 +0.61 % │     2912119046168     2899139910245 +0.45 % │     4627246846227     4626807738167 +0.01 % │    1075612    1070512  +0.48 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-coqprime │   80.43   79.83 +0.75 % │      221113740618      220058994910 +0.48 % │      309318064441      309257821138 +0.02 % │     768200     768404  -0.03 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴────────────────────────────────┴───────────────────┘
```

**Comment by gares** [comment]
I'm a bit uneasy with this in the 8.13+beta1 milestone and no review. Maybe @proux01 and @erikmd and help.

**Comment by proux01** [comment]
@gares I skimmed through all of it and it looks good. Let me do an actual review though.

@silene do you still want to target 8.13 ?

**Comment by silene** [comment]
> do you still want to target 8.13?

Sure.

> no review

Note that, while he did not leave any written comment, @bgregoir did review the 9 first commits, i.e., the ones that are not about primitive floating-point numbers.

**Comment by gares** [comment]
Great! It also needs an assignee, possibly from @coq/kernel-maintainers

**Comment by ppedrot** [comment]
I am assigning. I already reviewed superficially this PR some time ago (hence my comment above). I'll wait for further input and merge when @gares considers it adequate.

**Comment by gares** [comment]
Let's wait for @proux01 review

**Comment by gares** [comment]
@proux01 are you still working on the review? If not, it's still fine, but just tell us

**Comment by proux01** [comment]
@gares , sure, sorry for the delay, I'll try to complete it today.

**Review by proux01 — APPROVED** [review]
LGTM, just to be safe, as already discussed privately, could you add the following tests to `test-suite/primitive/float/next_up_down.v`

```Coq
Check (eq_refl : next_up nan = nan).
Check (eq_refl : next_down nan = nan).
Check (eq_refl : next_up neg_infinity = -0x1.fffffffffffffp+1023).
Check (eq_refl : next_down neg_infinity = neg_infinity).
Check (eq_refl : next_up (-0x1.fffffffffffffp+1023) = -0x1.ffffffffffffep+1023).
Check (eq_refl : next_down (-0x1.fffffffffffffp+1023) = neg_infinity).
Check (eq_refl : next_up (-0x1.ffffffffffffap+1023) = -0x1.ffffffffffff9p+1023).
Check (eq_refl : next_down (-0x1.ffffffffffffap+1023) = -0x1.ffffffffffffbp+1023).
Check (eq_refl : next_up (-0x1.fffffffffffff) = -0x1.ffffffffffffe).
Check (eq_refl : next_down (-0x1.fffffffffffff) = -0x1p+1).
Check (eq_refl : next_up (-0x1p1) = -0x1.fffffffffffff).
Check (eq_refl : next_down (-0x1p1) = -0x1.0000000000001p+1).
Check (eq_refl : next_up (-0x1p-1022) = -0x0.fffffffffffffp-1022).
Check (eq_refl : next_down (-0x1p-1022) = -0x1.0000000000001p-1022).
Check (eq_refl : next_up (-0x0.fffffffffffffp-1022) = -0x0.ffffffffffffep-1022).
Check (eq_refl : next_down (-0x0.fffffffffffffp-1022) = -0x1p-1022).
Check (eq_refl : next_up (-0x0.01p-1022) = -0x0.00fffffffffffp-1022).
Check (eq_refl : next_down (-0x0.01p-1022) = -0x0.0100000000001p-1022).
Check (eq_refl : next_up (-0x0.0000000000001p-1022) = -0).
Check (eq_refl : next_down (-0x0.0000000000001p-1022) = -0x0.0000000000002p-1022).
Check (eq_refl : next_up (-0) = 0x0.0000000000001p-1022).
Check (eq_refl : next_down (-0) = -0x0.0000000000001p-1022).
Check (eq_refl : next_up 0 = 0x0.0000000000001p-1022).
Check (eq_refl : next_down 0 = -0x0.0000000000001p-1022).
Check (eq_refl : next_up 0x0.0000000000001p-1022 = 0x0.0000000000002p-1022).
Check (eq_refl : next_down 0x0.0000000000001p-1022 = 0).
Check (eq_refl : next_up 0x0.01p-1022 = 0x0.0100000000001p-1022).
Check (eq_refl : next_down 0x0.01p-1022 = 0x0.00fffffffffffp-1022).
Check (eq_refl : next_up 0x0.fffffffffffffp-1022 = 0x1p-1022).
Check (eq_refl : next_down 0x0.fffffffffffffp-1022 = 0x0.ffffffffffffep-1022).
Check (eq_refl : next_up 0x1p-1022 = 0x1.0000000000001p-1022).
Check (eq_refl : next_down 0x1p-1022 = 0x0.fffffffffffffp-1022).
Check (eq_refl : next_up 0x1p1 = 0x1.0000000000001p+1).
Check (eq_refl : next_down 0x1p1 = 0x1.fffffffffffff).
Check (eq_refl : next_up 0x1.fffffffffffff = 0x1p+1).
Check (eq_refl : next_down 0x1.fffffffffffff = 0x1.ffffffffffffe).
Check (eq_refl : next_up 0x1.ffffffffffffap+1023 = 0x1.ffffffffffffbp+1023).
Check (eq_refl : next_down 0x1.ffffffffffffap+1023 = 0x1.ffffffffffff9p+1023).
Check (eq_refl : next_up 0x1.fffffffffffffp+1023 = infinity).
Check (eq_refl : next_down 0x1.fffffffffffffp+1023 = 0x1.ffffffffffffep+1023).
Check (eq_refl : next_up infinity = infinity).
Check (eq_refl : next_down infinity = 0x1.fffffffffffffp+1023).
```

**Comment by silene** [comment]
> just to be safe, as already discussed privately

We did, and then I looked at `test-suite/primitive/float/next_up_down.v`, and I concluded that the needed tests were already there. But you seem to think otherwise, so I must be missing something. For example, you are suggesting to add

```coq
Check (eq_refl : next_up infinity = infinity).
```

but there is already

```coq
Definition f6 := infinity.
Check (eq_refl : Prim2SF (next_up f6) = SF64succ (Prim2SF f6)).
```

Is it because you don't trust `SF64succ`?

**Comment by ppedrot** [comment]
It can't hurt to have more tests. I recently had trouble with this code which led me to #13274. So sandwiching definitions in the middle (or not) can reveal serious bugs in the kernel...

**Comment by proux01** [comment]
No specific reasons. Those tests are cheap and the parsing infrastructure now enables to write the constants directly, so why not.

**Comment by ppedrot** [comment]
Will merge when CI passes then.

**Comment by gares** [comment]
ci green, the stars are aligned

**Pull request merged**
